carbon-design-system/carbon
# feat(themes): add type tokens to themes

Closes #4104

Add support for type tokens in our themes. This is done through:

- Adding token reference in `@carbon/type` in `tokens.js` 
- Including `@carbon/type` in `@carbon/themes`
- Updating tests in `@carbon/themes` to reference tokens from `@carbon/type` in tests (will fail if they are not included in a theme
- Re-exporting default type tokens from `@carbon/type` in the `white` theme
- Re-exporting type tokens in other themes from `white` theme
- Updating build step to include all tokens
- Add filter step so that non-object token values are emitted in terms of CSS Custom properties (delegate map token values to type mixin)
- Add support for inlining CSS Custom Properties in the `type-style` mixin

#### Changelog

**New**

- `unstable_tokens` in `@carbon/type`
- Add support for node resolution for sass imports in tests in `test-utils/scss`

**Changed**

- Update `tokens` in `@carbon/themes` to include `type` in tests and themes
- Update build step to include type tokens
- Add support for `carbon--type-style` to emit CSS Custom Properties if `enable-css-custom-properties` is set to true

**Removed**

#### Testing / Reviewing

- Pull down PR
- Make sure local files are built correctly by running:

```bash
yarn lerna run build --scope=carbon-components --include-filtered-dependencies
```

- Run the vanilla demo by doing:

```bash
cd packages/components
yarn dev --use-custom-properties
```

- [ ] Verify that type tokens are being emitted

![image](https://user-images.githubusercontent.com/3901764/65793341-2f01c800-e12b-11e9-856e-6f9cafe5e76b.png)

- Update the selector below in `components/demo/scss/demo.scss` to include the following:

```scss
.demo--container,
:root.demo--theme--white .demo--container,
.component-example__live,
:root.demo--theme--white .component-example__live,
.bx--tooltip,
:root.demo--theme--white .bx--tooltip,
.bx--overflow-menu-options,
:root.demo--theme--white .bx--overflow-menu-options {
  @include carbon--theme(
    $theme:
      map-merge(
        $carbon--theme--white,
        (
          body-long-01: (
            font-size: 20px,
            font-weight: 400,
            line-height: 1.5rem,
            letter-spacing: 0,
          ),
        )
      ),
    $emit-custom-properties: true,
    $emit-difference: true
  );
}
```

- [ ] Verify type prescription has changed for `body-long-01`

![image](https://user-images.githubusercontent.com/3901764/65793606-dbdc4500-e12b-11e9-9333-579f9498ba44.png)

- [ ] Verify only CSS Custom property emitted is what has changed (body-long-01)

![image](https://user-images.githubusercontent.com/3901764/65793648-ee567e80-e12b-11e9-9900-1c65544ff903.png)

- [ ] Double-check that only CSS properties are being emitted in CSS Custom Properties (no random keys like breakpoints coming from type token maps)
